source/lmp/tests/test_lammps.py

```diff
@@ -0,0 +1,120 @@
+    ).split())
+
+
+
+@pytest.fixture
+def lammps() -> PyLammps:
+    lammps = PyLammps()
+    lammps.units("metal")
+    lammps.boundary("p p p")
+    lammps.atom_style("atomic")
+    lammps.neighbor("2.0 bin")
+    lammps.neigh_modify("every 10 delay 0 check no")
+    lammps.read_data(data_file.resolve())
+    lammps.mass("1 16")
+    lammps.mass("2 2")
+    lammps.timestep(0.0005)
+    lammps.fix("1 all nve")
+    yield lammps
+
+
+def test_pair_deepmd(lammps):
+    lammps.pair_style("deepmd {}".format(pb_file.resolve()))
+    lammps.pair_coeff("* *")
+    lammps.run(0)
+    assert lammps.eval("pe") == pytest.approx(expected_e)
+    for ii in range(6):
+        assert lammps.atoms[ii].force == pytest.approx(expected_f[ii])
+    lammps.run(1)
+
+
+def test_pair_deepmd_model_devi(lammps):
+    lammps.pair_style("deepmd {} {} out_file {} out_freq 1 atomic".format(pb_file.resolve(), pb_file2.resolve(), md_file.resolve()))
+    lammps.pair_coeff("* *")
+    lammps.run(0)
+    assert lammps.eval("pe") == pytest.approx(expected_e)
+    for ii in range(6):
+        assert lammps.atoms[ii].force == pytest.approx(expected_f[ii])
+    # load model devi
+    md = np.loadtxt(md_file.resolve())
+    expected_md_f = np.linalg.norm(np.std([expected_f, expected_f2], axis=0), axis=1)
+    assert md[7:] == pytest.approx(expected_md_f)
+    assert md[4] == pytest.approx(np.max(expected_md_f))
+    assert md[5] == pytest.approx(np.min(expected_md_f))
+    assert md[6] == pytest.approx(np.mean(expected_md_f))
+
+
+def test_pair_deepmd_model_devi_atomic_relative(lammps):
+    relative = 1.0
+    lammps.pair_style("deepmd {} {} out_file {} out_freq 1 atomic relative {}".format(pb_file.resolve(), pb_file2.resolve(), md_file.resolve(), relative))
+    lammps.pair_coeff("* *")
+    lammps.run(0)
+    assert lammps.eval("pe") == pytest.approx(expected_e)
+    for ii in range(6):
+        assert lammps.atoms[ii].force == pytest.approx(expected_f[ii])
+    # load model devi
+    md = np.loadtxt(md_file.resolve())
+    norm = np.linalg.norm(np.mean([expected_f, expected_f2], axis=0), axis=1)
+    expected_md_f = np.linalg.norm(np.std([expected_f, expected_f2], axis=0), axis=1)
+    expected_md_f /= norm + relative
+    assert md[7:] == pytest.approx(expected_md_f)
+    assert md[4] == pytest.approx(np.max(expected_md_f))
+    assert md[5] == pytest.approx(np.min(expected_md_f))
+    assert md[6] == pytest.approx(np.mean(expected_md_f))
+
+
+def test_pair_deepmd_model_devi_atomic_relative_v(lammps):
+    relative = 1.0
+    lammps.pair_style("deepmd {} {} out_file {} out_freq 1 atomic relative_v {}".format(pb_file.resolve(), pb_file2.resolve(), md_file.resolve(), relative))
+    lammps.pair_coeff("* *")
+    lammps.run(0)
+    assert lammps.eval("pe") == pytest.approx(expected_e)
+    for ii in range(6):
+        assert lammps.atoms[ii].force == pytest.approx(expected_f[ii])
+    #md = np.loadtxt(md_file.resolve())
+    # TODO: how to get the virial?
```

Comment:
What I did is `compute pressure` and then subtract the kinetic contribution.

Reply:
Another option may be [using `compute stress/atom NULL virial`](https://docs.lammps.org/compute_stress_atom.html), but I have never tried.

Reply:
I got all zero values from `virial all stress/atom NULL virial`. There may be something else wrong.
```py
    lammps.compute("virial all stress/atom NULL virial")
    lammps.dump("1 all custom 1 dump id c_virial[1] c_virial[2] c_virial[3]")
```

```dump
ITEM: TIMESTEP
0
ITEM: NUMBER OF ATOMS
6
ITEM: BOX BOUNDS xy xz yz pp pp pp
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
ITEM: ATOMS id c_virial[1] c_virial[2] c_virial[3]
1 -0 -0 -0
2 -0 -0 -0
3 -0 -0 -0
4 -0 -0 -0
5 -0 -0 -0
6 -0 -0 -0
```

Reply:
I see. `compute centroid/stress/atom` should be used instead...

Reply:
Really weird, I got
```
ITEM: TIMESTEP
0
ITEM: NUMBER OF ATOMS
6
ITEM: BOX BOUNDS xy xz yz pp pp pp
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
0.0000000000000000e+00 1.3000000000000000e+01 0.0000000000000000e+00
ITEM: ATOMS id c_virial[1] c_virial[2] c_virial[3] c_virial[4] c_virial[5] c_virial[6] c_virial[7] c_virial[8] c_virial[9]
1 -466591 -505066 -414063 -80970.2 452837 166487 -60892.5 444864 169974
2 -500155 -61207.7 -339939 66365 427598 -103312 135917 404495 -43696.3
3 -138177 -319263 -235570 -205665 224036 289084 -289945 245113 225895
4 -415602 -160778 -376371 -66875.1 54457.6 243892 -67876.8 54255.9 244443
5 1524.56 -153220 -381473 965.674 -12397.8 240780 -26327.3 31935.1 242059
6 -370016 -54260 -97381.6 -129762 161594 73006.7 -106816 127462 71263
```

Reply:
The unit is different..